front/src/modules/simulationResult/components/SpaceTimeChart/useLazyProjectTrains.ts

```diff
@@ -0,0 +1,158 @@
+import { useEffect, useState, type Dispatch, type SetStateAction, useMemo } from 'react';
+
+import { keyBy } from 'lodash';
+import { useSelector } from 'react-redux';
+
+import upsertNewProjectedTrains from 'applications/operationalStudies/helpers/upsetNewProjectedTrains';
+import type { TrainSpaceTimeData } from 'applications/operationalStudies/types';
+import {
+  osrdEditoastApi,
+  type PathfindingResultSuccess,
+  type TrainScheduleResult,
+} from 'common/api/osrdEditoastApi';
+import { useOsrdConfSelectors } from 'common/osrdContext';
+import type { Dictionary } from 'utils/types';
+
+import { getTrainIdsPackage } from './utils';
+
+const BATCH_SIZE = 10;
+
+type useLazyLoadTrainsProp = {
+  infraId?: number;
+  trainIdsToProject: number[];
+  path?: PathfindingResultSuccess;
+  trainSchedules?: TrainScheduleResult[];
+  moreTrainsToCome?: boolean;
+  setTrainIdsToProject: Dispatch<SetStateAction<number[]>>;
+};
+
+/**
+ * This hook gradually projects trains of the timetable.
+ *
+ * It projects the trains by batch of 10 on the path.
+ * This optimizes the performance of the application and allow us to display the
+ * trains on the space time chart as soon as they are ready.
+ */
+const useLazyProjectTrains = ({
+  infraId,
+  trainIdsToProject,
+  path,
+  trainSchedules,
+  moreTrainsToCome = false,
+  setTrainIdsToProject,
+}: useLazyLoadTrainsProp) => {
+  const { getElectricalProfileSetId } = useOsrdConfSelectors();
+  const electricalProfileSetId = useSelector(getElectricalProfileSetId);
+
+  const [projectedTrainsById, setProjectedTrainsById] = useState<Dictionary<TrainSpaceTimeData>>(
+    {}
+  );
+  const projectedTrains = useMemo(() => Object.values(projectedTrainsById), [projectedTrainsById]);
+
+  const [nextPackageToProjectIndex, setNextPackageToProjectIndex] = useState(0);
+  const [readyToProjectNextPackage, setReadyToProjectNextPackage] = useState(true);
```

Comment:
It seems like this piece of state is just used to trigger the callback to project the next 10 trains. Can't we use a callback instead of state?

```ts
const projectNextTrains = async () => {
  try {
    // ...
  } finally {
    projectNextTrains();
  }
};
```

Reply:
Done, with the refacto of useLazyProjectTrains